### ISSUE TITLE
Support creating variables that are not logged even on TRACE level

### ISSUE THREAD

**Issue by pekkaklarck** [robotframework/robotframework#4537]
Often automation we need to use passwords or other values that should not be logged. To help with that, various libraries have keywords that avoid logging given values even though their keywords normally do that. For example, SeleniumLibrary has `Input Password` that doesn't log the password and also temporarily disables Robot's logging to avoid the password being logged by the underlying Selenium as well. This works fine otherwise, but if tests are run with `--loglevel TRACE`, Robot anyway logs all argument values and the password will be visible. Browser's `Fill Secret` avoids that problem by requiring the value to be passed using a special `$var` syntax so that Robot won't log the actual variable value. Unfortunately that doesn't work if the value is passed to `Fill Secret` via a user keyword and only that keyword uses the `$var` syntax like in the following example:

```robotframework
*** Variables ***
${PASSWORD}        secret

*** Test Cases ***
Example
    Keyword    ${PASSWORD}    # This value is logged on TRACE level

*** Keywords ***
Keyword
    [Arguments]    ${pwd}
    Fill Secret    selector    $pwd
```

To make it easier to use values that should not be logged, I propose we add new variable syntax `*{var}` to create "secret variables". The syntax could be used when creating variables in the Variables section, when creating local variables based on keyword return values, and with `Set Global/Suite/Test/Local Variable` keywords. The syntax would be only used for creating variables, they would be used normally like `${var}`:

```robotframework
*** Variables ***
*{PASSWORD}        secret

*** Test Cases ***
Example
    Keyword    ${PASSWORD}
    *{local} =    Another keyword
    Keyword    ${local}

*** Keywords ***
Keyword
    [Arguments]    ${pwd}
    Fill Secret    selector    ${pwd}
```

Technically the `*{var}` assignment would create a custom object with this kind of implementation:

```python
class Secret:
    def __init__(self, value, name=None):
        self.value = value
        self.name = name
    def __str__(self):
        return f'<Secret {self.name}>' if self.name else '<Secret>'
```

These objects would be used when the variable is passed around between test and user keywords. The real value would, however, be passed to library keywords so they would get values normally and wouldn't need to be changed to benefit from this new syntax.

Although this new syntax would work with existing keywords, we should make it possible for library keywords to require arguments to be "secret". That's basically what `Fill Secret` does but this new syntax would avoid the problem that the value can be logged on higher level. A convenient way to support this would be type hints:

```python
from robot.api.secrets import Secret

def example(secret: Secret):
    ...
```

This would work so that Robot would validate that the used value is a `Secret` and fail if it's not. I believe we should pass the real value, not the actual `Secret` object, to the underlying keyword also in this case. That can confuse type checkers but keywords can use [typing.cast](https://docs.python.org/3/library/typing.html#typing.cast) if that's a problem. To support normal type conversion, we should also allow parameterizing the type hint like `Secret[int]`.

Keywords could also return secret values simply by returning `Secret` instances:

```python
from robot.api.secrets import Secret

def example():
    return Secret('value')
```

**Comment by pekkaklarck** [comment]
One problem with the proposed implementation is that although Robot itself wouldn't log secret variable values, keywords could accidentally log them. We could easily special case keywords like `Log`, but because other keywords could leak the value that alone probably doesn't make sense. One possibility would be making it an error to pass a `Secret` to a library keyword that doesn't explicitly require `Secret`. An obvious problem of this approach is that secret variables wouldn't work with existing keywords. I got a feeling that it is a too big problem and we shouldn't to this. Possibly this behavior could be configurable, though, but in that case I'd add configuration only later.

If we'd made using secrets more script like this, I believe we should allow users to use `${secret.value}` syntax to access the actual value. This would make it possible to log the value for debug purposes. We should also allow keywords to accept also other values than secrets like `arg: Secret | str`.

**Comment by pekkaklarck** [comment]
One open question is what to do if a secret variable is added into a `@{list}` or `&{dict}` in the Variables section:
```robotframework
*** Variables ***
*{SECRET}        value
@{LIST}    ${SECRET}
&{DICT}    key=${SECRET}
```
I believe we should add the `Secret` object, not the value, to containers in cases like this. Otherwise if these variables are logged on TRACE, the secret value is logged as part of them. If the variables are passed to keywords like `Keyword    ${LIST}`, I don't think it makes sense for us to go through their contents and convert `Secret` objects to values like the plan is to do when only a secret value is used. These values can be deeply nested and also recursive making going through them hard. Not unwrapping `Secret` objects means keywords can encounter these objects, but I expect that to be rare and accessing the actual value is anyway easy.

If collections are created dynamically in tests or keywords like
```
${list} =    Create List    ${SECRET}
```
they end up getting the real value, not `Secret` object, because that's what is planned to be passed to `Create List`. The comment above discusses other alternatives to these plans, though.

**Comment by pekkaklarck** [comment]
Another questions is should we allow using secrets as parts of strings like `Hello, ${SECRET}!`. I believe this should be prohibited because the whole `Hello, ${SECRET}!` argument may be logged using TRACE after the variable is resolved. Using `Hello, ${SECRET.value}!` would work, though, unless we hide the actual value somewhere (which is rather complicated with Python).

**Comment by pekkaklarck** [comment]
Yet another question is deciding what to do if variables are _used_ like `*{var}`:
```
Keyword    *{SECRET}
```
Options include:
- Allow using exactly like `${SECRET}`. I don't like this too much. Having two alternative ways to do same thing is seldom a good idea.
- Come up some use case for. It could, for example, mean unwrapping `Secret´ and normal `${SECRET}` usage would pass the `Secret` object, not the value. Or possibly it could mean passing `Secret` while `${SECRET}` would unwrap.
- Make it an error now possibly come up with some usage for the syntax later.

**Comment by pekkaklarck** [comment]
Issue #4674 proposes supporting `$var` syntax as an alternative to `${var}`. If we do that, passing secrets to the Browser library by using the `$var` syntax won't work anymore. That makes the priority of this issue higher.

**Comment by Snooz82** [comment]
I would propose to NOT implement `*{secret}` because it is pretty implicit and only some super nerds will know that feature.

i would propose to implement type converters for variables and user keyword arguments first! Then Secret creation is for free.


```
*** Variables ***
${secret}:Secret    123456
```

https://github.com/robotframework/robotframework/issues/3278

discussion about that:
https://robotframework.slack.com/archives/C04FPJ3R4PN/p1678217116430379?thread_ts=1678198681.294349&channel=C04FPJ3R4PN&message_ts=1678217116.430379

**Comment by pekkaklarck** [comment]
Using `${pwd}:Secret` (or `${pwd:Secret}`, depending on how #3278 is implemented) would be a good alternative to `*{pwd}`. One downside is that variable type conversion is harder to implement than adding `*{pwd}` syntax, but that feature also has lot of other usages so implementing it makes sense anyway. If we go that route, then this issue obviously is dependent on #3278.

If we decide to rely on variable conversion instead of adding `*{var}` syntax, we need to add a custom variable converter for `Secret` that converts the value to a `Secret` object as discussed above. Then we need another argument converter for libraries that only validates that the given argument is a `Secret`.

**Comment by pekkaklarck** [comment]
Here's some proto code how `Secret` could possibly look like. It supports also `Secret[int]` and converts value accordingly. If this is added to Robot, values naturally need to be converted using Robot's converters.

```python
class Secret:
    type = None

    def __init__(self, value, name=None):
        self.__value = value
        self.name = name

    @property
    def value(self):
        return self.type(self.__value) if self.type else self.__value

    @classmethod
    def __class_getitem__(cls, item):
        return type(f'{cls.__name__}[{item.__name__}]', (cls,), {'type': item})

    def __str__(self):
        name = type(self).__name__
        args = 'value=<secret>'
        if self.name:
            args += f', name={self.name!r}'
        return f'{name}({args})'
```

**Comment by pekkaklarck** [comment]
Let's postpone this until we have made decision about variable type conversion (#3278) and can we possibly use that approach with secret variables.

**Comment by sebastianciupinski** [comment]
@pekkaklarck you should not think about how to hide secrets in a way you do. hide it in log is enough, but do not think if using this in concatenation shall be allowed or not, etc. User can always: convert secret to bytes, write secret to file, etc.

**Comment by pekkaklarck** [comment]
We prototyped variable type conversion (#3278) and it seems getting it done in RF 7.3 is not too complicated. After that we can look can we get this feature implemented using the proposed `${var: Secret}` approach.

**Comment by mkorpela** [comment]
To make the Secret type really work you can’t make non-secrets secrets and you must protect secrets from
becomming anything else.
That feels to me that it will be pretty limiting on where this type can be actually used.

**Comment by aaltat** [comment]
I have two needs: 

1. Prevent the secret from being visible in RF logs 
2. Prevent secret to be present in Robot Framework test data as clear text

Python side hardening or adding some magic in log files are not part of this feature request.

**Comment by JFoederer** [comment]
1 is clear.
For 2 it is the question whether you need to actively block that option, or that it is better to offer that possibility as a separate feature and leave it up to the user what they use.

I think the latter makes the scope much sharper, with less open questions on how deep the secret protection goes.
To rephrase:

1. _A secret variable can be assigned without it being present in readable form anywhere._
2. _Secret variables are blanked out when logged_

**Comment by JFoederer** [comment]
That is a very nice goal @fkromer! But we can't expect Robot Framework to cover all that is needed for that off-the-shelf by Release 7.4. What is achievable, is providing a tool with which experts in that area can already achieve their goal. That is why we need to be very clear on what _is_ and what _is not_ in scope now.

**Comment by joseajunior** [comment]
This idea is fascinating, but couldn't we improve on readability?
Instead of using: `--variable "PASSWORD: Secret:123456789"`, couldn't we create some `--secret` option and call something like:
`robot --secret "PASSWORD:123456789" .`
That's my two cents on this.

**Comment by aaltat** [comment]
>  couldn't we improve on readability? Instead of using: `--variable "PASSWORD: Secret:123456789"`, couldn't we create some `--secret` option and call something like: `robot --secret "PASSWORD:123456789" .` 

This feature is based on the [variable type conversion](https://github.com/robotframework/robotframework/blob/master/doc/releasenotes/rf-7.3rc3.rst#variable-type-conversion) on Robot Framework 7.3. The 7.3 release already support conversion from command line, like example `--variable "NUMBER: int:42"`. This enhancement is adding new type to the [supported conversions](https://robotframework.org/robotframework/latest/RobotFrameworkUserGuide.html#supported-conversions) table, with some [additional features](https://github.com/robotframework/robotframework/issues/4537#issuecomment-2903601231) explained above. Although I like readability, which your proposal offers, it is also brings more complexity, because now there would be two ways to create secrets. And as the [Python Zen](https://peps.python.org/pep-0020/) says: `There should be one-- and preferably only one --obvious way to do it.`

**Comment by aaltat** [comment]
> To make the Secret type really work you can’t make non-secrets secrets and you must protect secrets from becomming anything else. That feels to me that it will be pretty limiting on where this type can be actually used.

Yes your point is correct and it is limiting. But it is done by a purpose by two reasons

1. It is easier to be more limiting now and allow more access later than other way around
2. If we allow people to expose secret to converted in RF test data to other type. People will accidentally do it. 

For first point, there is no need to move use to secret if you do not want to. One can pass around secret as normal variables like now. Type hint is also optional, if there is no type hint, no conversion is made and user can pass the variable around as he/she wants.  Also I did not plan to force people to use secrets in Browser library keywords for foreseen future, string will be accepted like now (disclaimer, I do not see very far in the future either.)

For the second point, I have seen many times in people test data something like this: 
```robotframework
*** Variables ***
${PASSWORD}    123456789
```
Where the password is actually their user account password and the test data stored in version control and can be read by anyone in the company. The need is that we want to limit the ways people can shoot them self in the foot. But because this is light weight secret handling, there will be ways to get the secrets as clear text. The current PR, the secret is stored in a class as plain text. So if you know in which attribute the secret is stored, then you log it in plain text also in RF test data.

**Comment by aaltat** [comment]
> One use case is automated _Security Configuration Assessment_ (SCA). Surprisingly there are little RoboCon presentations about security related test automation in general. But actually it's super interesting.

I do agree and I hope that there would be such presentations, but SCA not the aim of this feature. The aim for 7.4 release is 

1. Prevent the secret from being visible in RF logs
2. Prevent secret to be present in Robot Framework test data as clear text

That being said, I hope secret handling would get more traction also in RF community.

**Comment by JFoederer** [comment]
The only part where our opinions differ is in whether it must be mandatory to set secret variables from an already secret source, or that non-secret sources may be used as well.

I think this comes down to usability. In your vision @aaltat, what is the simplest way to set a secret variable? If I am only interested in keeping the passwords out of the logging, what extra effort is involved to already keep it secret during assignment? If this would become mandatory?

**Comment by aaltat** [comment]
> The only part where our opinions differ is in whether it must be mandatory to set secret variables from an already secret source, or that non-secret sources may be used as well.

This is very good question and answer depends. But when looking this from Browser library point of view, it will not demand of using secrets in Type/Input Secret keywords in the foreseen future (but I do not see far in the future either and some decisions are not in my hand either.) But I only maintain that library and other library maintainers might decide something else.

> I think this comes down to usability. In your vision [@aaltat](https://github.com/aaltat), what is the simplest way to set a secret variable? If I am only interested in keeping the passwords out of the logging, what extra effort is involved to already keep it secret during assignment? If this would become mandatory?

Good question and again answer depends. Perhaps the simplest way is to set it from command line: `robot --variable "PASSWORD: Secret:123456789"`, or reading from environment variable. But example in GitHub I would at least using [secrets](https://docs.github.com/en/actions/security-for-github-actions/security-guides/using-secrets-in-github-actions) (many similar system exist in other CI systems too) but this is very environment specific question and there are solutions like [Azure key vault](https://azure.microsoft.com/en-us/products/key-vault) and to other custom build solutions available. In my current job, there is not one single correct way to handle secrets (because of the long history of the company and many different ways to build a software.) and we at my work need to support many ways. 

The point being, these are good question but answer depends on context. The aim so to make this flexible (well, flexible is relative term when talking about security) and optional for users. This enhancement does not force anyone to use secrets.

**Comment by pekkaklarck** [comment]
Two clarifications to @aaltat 's excellent feature specification:

1. Using `VAR    ${x: Secret}    ${y}`, and similar syntax in the Variables section, will work _if_ `${y}` is Secret.

2. `${x: Secret}=    Keyword ` will _only_ work if the keyword returns a Secret. No automatic conversion like with normal types.

**Comment by pekkaklarck** [comment]
PR #5449 by @aaltat implementing this functionality has been merged, but there are some things still fix and enhance:

- [x] Deciding where the `Secret` type implementation should be.
- [x] Fixing corner cases related to creating lists/dicts with lists/dicts in their values.
- [x] Fixing smallish issues explained in the inline comments of the final review of PR #5449.
- [x] Reviewing documentation. I looked at the docs under `src` but haven't looked at the actual User Guide docs at all. I want to get the implementation ready (and possibly even released as an alpha) before that.

**Comment by pekkaklarck** [comment]
I tried moving the `Secret` implementation under `robot.api.types`, but the cyclic import issues turned out to be too hard to fix. We've many times earlier exposed classes, functions, etc. implemented elsewhere via `robot.api`, so I guess that's not a major issue in this case either. The implementation is currently in `robot.utils._secret` module, but I don't think trying to hide it into a private module has benefits, especially because it is anyway public exposed via `robot.utils`. I'll thus rename the module to `robot.utils.secret`. The class also needs a note that it should be imported via `robot.api.types` and that module itself needs docs too.

**Comment by pekkaklarck** [comment]
The new [robot.api.types](https://robot-framework.readthedocs.io/en/master/autodoc/robot.api.html#module-robot.api.types) module now has API docs and those docs link to the `Secret` class. There were some technical challenges with that due to #5498.

**Comment by pekkaklarck** [comment]
Above commits finalized documentation so this issues is now done! 🎉
